### PR TITLE
feat: Add strict_mode option to function_schema and function_tool

### LINKED ISSUE #43: example @function_tool with optional parameters
I'm struggling to find a way to declare a function which has an optional parameter. Ideally, I want something like this. Let me know what's the way, if you don't mind!

```python

@function_tool
def get_latest_elasticsearch_version(major_version: int | None = None) -> str:
    """Returns the latest GA version of Elasticsearch in "X.Y.Z" format.

    Args:
        major_version: Major version to filter by (e.g. 7, 8). Defaults to latest
    """

--snip--
```

### PR DESCRIPTION
**Pull request by Jai0401**
This PR introduces a `strict_mode: bool = True` option to `@function_tool`, allowing optional parameters when set to False. This change enables more flexibility while maintaining strict JSON schema validation by default.

resolves #43 

## Changes:

- Added `strict_mode` parameter to `@function_tool` and passed it to `function_schema` and `FunctionTool`.
- Updated `function_schema.py` to respect `strict_mode` and allow optional parameters when set to False.
- Added unit tests to verify optional parameters work correctly, including multiple optional params with different types.

## Tests:

- Verified function calls with missing optional parameters behave as expected.
- Added async tests to validate behavior under different configurations.

### CODE REVIEW

**Comment by Jai0401** [comment]
Hey @codefromthecrypt, I used your suggested edge cases (e.g., multiple optional parameters of different types) as inspiration for the test cases in this PR. Thanks for the great examples! Let me know if you have any additional suggestions or feedback.

**Comment by Jai0401** [comment]
Hey @rm-openai, I've implemented the strict_mode option for `@function_tool()` and added test cases inspired by @codefromthecrypt's suggestions. Could you review the PR when you get a chance? Let me know if any changes are needed. Thanks!

File: src/agents/tool.py [hunk 1]
```diff
@@ -186,6 +189,7 @@ def function_tool(
         failure_error_function: If provided, use this function to generate an error message when
             the tool call fails. The error message is sent to the LLM. If you pass None, then no
             error message will be sent and instead an Exception will be raised.
+        strict_mode: If False, allows optional parameters in the function schema.
```
Comment (codefromthecrypt) [inline src/agents/tool.py]:
```suggestion
        strict_mode: If False, parameters with default values become optional in the function schema.
```

**Review by codefromthecrypt — APPROVED** [review]
tests look easier than I thought. you found a nice place to put them!

File: src/agents/tool.py (hunk 1 above)
Comment (Jai0401) [reply src/agents/tool.py]: src/agents/tool.py:192:101: E501 Line too long (101 > 100)
linting error

Comment (Jai0401) [reply src/agents/tool.py]: Fixed linting errors while preserving changes

Comment (codefromthecrypt) [reply src/agents/tool.py]: thanks!

**Review by rm-openai — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
@rm-openai if you don't mind cutting a release by weds, that would be great. I'm using this as a demo at the red hat user group in Cary, NC thursday https://www.meetup.com/raleigh-rhug/events/306421516/?eventOrigin=group_upcoming_events
